samples/DevTestLabs/Modules/Library/Az.DevTestLabs2.psm1

```diff
@@ -0,0 +1,2636 @@
+      }
+    } catch {
+      Write-Error -ErrorRecord $_ -EA $callerEA
+    }
+  }
+  end {}
+}
+
+function Set-AzDtlVmAutoStart {
+  [CmdletBinding()]
+  param(
+    [parameter(Mandatory=$true,HelpMessage="Vm to apply autostart to", ValueFromPipeline=$true)]
+    [ValidateNotNullOrEmpty()]
+    $Vm,
+
+    [parameter(Mandatory=$false, ValueFromPipelineByPropertyName = $true,HelpMessage="Name of the repository to get artifact from.")]
+    [bool]
+    $AutoStartStatus = $true
+  )
+
+  begin {. BeginPreamble}
+  process {
+    try {
+      foreach($v in $Vm) {
+        Write-Verbose "Set AutoStart for $($v.ResourceId) to $AutoStartStatus"
+        Set-AzureRmResource -ResourceId $v.ResourceId -Tag @{AutoStartOn = $AutoStartStatus} -Force | Out-Null
+        $v  | Get-AzDtlVm
+      }
+    } catch {
+      Write-Error -ErrorRecord $_ -EA $callerEA
+    }
+  }
+  end {}
+}
+
+$templateVmCreation = @"
+{
+  "`$schema": "http://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+  "contentVersion": "1.0.0.0",
+  "parameters": {
+    "newVMName": {
+      "type": "string"
+    },
+    "existingLabName": {
+      "type": "string"
+    }
+  },
+  "variables": {
+    "labSubnetName": "[concat(variables('labVirtualNetworkName'), 'Subnet')]",
+    "labVirtualNetworkId": "[resourceId('Microsoft.DevTestLab/labs/virtualnetworks', parameters('existingLabName'), variables('labVirtualNetworkName'))]",
+    "labVirtualNetworkName": "[concat('Dtl', parameters('existingLabName'))]",
+    "resourceName": "[concat(parameters('existingLabName'), '/', parameters('newVMName'))]",
+    "resourceType": "Microsoft.DevTestLab/labs/virtualMachines"
+  },
+  "resources": [
+    {
+      "apiVersion": "2018-10-15-preview",
+      "type": "Microsoft.DevTestLab/labs/virtualMachines",
+      "name": "[variables('resourceName')]",
+      "location": "[resourceGroup().location]",
+      "properties": {
+        "labVirtualNetworkId": "[variables('labVirtualNetworkId')]",
+        "labSubnetName": "[variables('labSubnetName')]"
+      }
+    }
+  ],
+  "outputs": {
+    "vmId": {
+      "type": "string",
+      "value": "[resourceId(variables('resourceType'), parameters('existingLabName'), parameters('newVMName'))]"
+    }
+  }
+}
+"@
+
+# TODO: Consider adding parameters Name and ResourceGroupName tied to the pipeline by property
+# to enable easier pipeline for csv files in the form [Name, ResourceGroupName, VMName, ...]
+function New-AzDtlVm {
+  [CmdletBinding()]
+  [Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSAvoidUsingPlainTextForPassword", "", Scope="Function")]
+  param(
+    [parameter(Mandatory=$false, ValueFromPipeline=$true, HelpMessage="Lab object to create Vm into. This is not used if you specify Name/ResourceGroupName")]
+    [ValidateNotNullOrEmpty()]
+    $Lab,
+
+    [parameter(Mandatory=$false, ValueFromPipelineByPropertyName = $true, HelpMessage="Name of Lab object to create Vm into")]
+    [ValidateNotNullOrEmpty()]
+    $Name,
```

Comment:
Agreed, but this is an unfortunate effect (the only one to my knowledge for this library) of using the pipeline to propagate state as sometimes the state needs to be in a particularly named property. Changing this name would break pipelines later on. I suggest leaving it as it is for now as a small inconvenience and revisit if people get confused by it.